Documentation improvements

Warchant opened the pull request:
- [x] add LICENSE
- [x] add code of conduct
- [x] improve github PR template
- [x] add changelog generator config
- [x] improve CONTRIBUTING.md
- [ ] generate a CHANGELOG.md (it is autogenerated with https://github.com/skywinder/github-changelog-generator)

TODO: review and fix https://github.com/hyperledger/iroha/wiki/Iroha-working-agreement
But it will be done in next PRs.

neewy commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,46 @@
+
+## Our Pledge
+
+In the interest of fostering an open and welcoming environment, we as contributors and maintainers pledge to making participation in our project and our community a harassment-free experience for everyone, regardless of age, body size, disability, ethnicity, gender identity and expression, level of experience, nationality, personal appearance, race, religion, or sexual identity and orientation.
```
nice!

neewy commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,46 @@
+
+Project maintainers are responsible for clarifying the standards of acceptable behavior and are expected to take appropriate and fair corrective action in response to any instances of unacceptable behavior.
+
+Project maintainers have the right and responsibility to remove, edit, or reject comments, commits, code, wiki edits, issues, and other contributions that are not aligned to this Code of Conduct, or to ban temporarily or permanently any contributor for other behaviors that they deem inappropriate, threatening, offensive, or harmful.
```
Do we need to tell about exact violation when we have to fix violations?

Warchant replied:
It is exact copy of https://www.contributor-covenant.org . I guess, it should be here.

neewy commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,46 @@
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team [here](./CONTRIBUTING.md/?#contact-developers). All complaints will be reviewed and investigated and will result in a response that is deemed necessary and appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an in ...
```
This is good

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+
+A *question* is any discussion that is typically neigher a bug, nor feature requests, nor improvements - "How do I do X?".
+
+Questions are tracked as [Github Issues](https://guides.github.com/features/issues/) or via private messages in  [your favourite messenger](#contace-developers).
```
contact, not contace

neewy replied:
after IN there are two spaces, please remove one

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+
+- **Write tests** for new code.
+
+- Every pull request should be reviewed and **get at least two approves**.
```
approvals

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+
+- Include issue numbers in Pull Request body only.
+
+- When finished work, **rebase onto base branch** with 
```
What if we will add guide?

Warchant replied:
Okay, I will post the link.

Warchant replied:
*fixed*

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+- **Use present tense** ("Add feautre", not "Added feature").
+- **Use imperative mood** ("Deploy docker to..." not "Deploys docker to...").
+- Write meaningful commit message.
+- **Sign every commit** with [DCO](https://github.com/apps/dco): `Signed-off-by: $NAME <$EMAIL>`. 
```
Maybe add guide what to do if contributor messed up with signing?

Warchant replied:
Good point.

Warchant replied:
There is actually link to DCO, it has tutorial and description.

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+- Use [clang-format](http://clang.llvm.org/docs/ClangFormat.html) for code formatting. 
+- Follow [CppCoreGuidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) and [Cpp Best Practices](https://lefticus.gitbooks.io/cpp-best-practices).
+- Avoid platform-dependent code.
+- Use C++14.
```
Maybe link to standard?

Warchant replied:
Good point.

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+
+- Use [clang-format](http://clang.llvm.org/docs/ClangFormat.html) for code formatting. 
+- Follow [CppCoreGuidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) and [Cpp Best Practices](https://lefticus.gitbooks.io/cpp-best-practices).
+- Avoid platform-dependent code.
```
Maybe example?

Warchant replied:
If somebody come up with good example, I will include the link.

neewy replied:
you write platform-dependent code, but you don't introduce the definition or make any link. I am sad

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+| `feature`               | Feature requests.                        |
+| `good-first-issue`      | Good starting point to begin contributing. |
+| `help-wanted`           | Maintainers ask for help to work on this issue. |
+| `needs-reproduction`    | Issue is likely a bug, but needs to be reproduced and confirmed. |
```
nice

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,197 @@
+
+Thank you for reading whole document! 
+
+Your code phrase is "I promise to follow Contributing Guidelines."* without quotes.
```
cool

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,28 @@
+
+<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->
+
+### Alternate Designs
```
Probably I have to mention rules about optional sections.

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,28 @@
+
+### Applicable Issues
+
+<!-- Enter any applicable Issues here -->
```
It is not about github issues, it is about issues with implementation.
Example: "I implemented CLI, but there may be undefined behaviour if user use START before CREATE".

Warchant commented on CHANGELOG.md:
```diff
@@ -0,0 +1,552 @@
+# Change Log
```
It is autogenerated, can be pasted into release notes, so why not?

Warchant replied:
GitIssueBot will enforce users to write clear and descriptive titles. It will help significantly during changelog generation.

neewy replied:
Actually I agree

Warchant replied:
As soon as our names of PRs and issues will be good and descriptive, this changelog will be very valuable. I don't think it is a pollution of a repository.

neewy replied:
It is a file with 1000 of lines. Change log should be generated for each release, not for whole repo. Please delete it.

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+
+### Git Style Guide
+
+- **Use present tense** ("Add feautre", not "Added feature").
```
"instead of" will be useful here

neewy replied:
"Add feautre" -- feature

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+- Limit the first line of commit message to 50 characters or less.
+- When only changing documentation, include `[ci skip]` in the commit description.
+- We use mixed approach of [Github Flow](https://guides.github.com/introduction/flow/) and [Git Flow](http://nvie.com/posts/a-successful-git-branching-model/). More at [Iroha Working Agreement](https://github.com/hyperledger/iroha/wiki/Iroha-working-agreement#2-version-control-system).
+
```
Adding a blank line after first one in the commit message is also a good practice

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+- Use [clang-format](http://clang.llvm.org/docs/ClangFormat.html) for code formatting. 
+- Follow [CppCoreGuidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) and [Cpp Best Practices](https://lefticus.gitbooks.io/cpp-best-practices).
+- Avoid platform-dependent code.
+- Use [C++14](https://en.wikipedia.org/wiki/C%2B%2B14).
```
s/Use/Prefer/

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+- Follow [CppCoreGuidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) and [Cpp Best Practices](https://lefticus.gitbooks.io/cpp-best-practices).
+- Avoid platform-dependent code.
+- Use [C++14](https://en.wikipedia.org/wiki/C%2B%2B14).
+- Use [camelCase](https://en.wikipedia.org/wiki/Camel_case).
```
wtf D: better add link to google c++ code style

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+
+| Label Name              | Description                              |
+| :---------------------- | ---------------------------------------- |
+| `enhancement:code`      | Any improvements in **existing** code.   |
```
Markdown supports html (hopefully github's one too) so place html of that issue, e.g:
```<a href="/hyperledger/iroha/labels/enhancement" class="label css-truncate-target linked-labelstyle-1d76db" title="enhancement">enhancement</a>```
I guess you know where to find it

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+
+## Additional Notes
+
+### Issue and Pull Request Labels
```
Maybe it's better separate it? E.g `question` and `good-first-issue` are only applicable to issues

l4l replied:
As I see all of the labels applicable to question, so it's possible to make a `PR labels` section and `other labels` section

l4l replied:
Also it's better place it in other file, usually labels is set by maintainers

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `status:inactive`       | Inactive PR or Issue. Likely to become a `candidate-for-closing` |
+| `status:pending`        | Issue is posted, but is not reviewed by maintainers. |
+| `status:resolved`       | Resolved issue.                          |
+| `status:wontfix`        | Core team has decided not to fix these issue for now. |
```
"won't" relates to future. Why is "for now"? Remove it plz

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `status:in-progress`    | Work in progress.                        |
+| `status:inactive`       | Inactive PR or Issue. Likely to become a `candidate-for-closing` |
+| `status:pending`        | Issue is posted, but is not reviewed by maintainers. |
+| `status:resolved`       | Resolved issue.                          |
```
What's the difference with the closed one?

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `pri:normal`            | Normal priority.                         |
+| `pri:important`         | Important issue.                         |
+| `pri:critical`          | Critical issue. Must be fixed immediately. |
+| `pri:blocker`           | Issue blocked on other issues.           |
```
s/on/by/

l4l replied:
Also don't really sure that we need this one, critical imo fine for now

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `pri:blocker`           | Issue blocked on other issues.           |
+| `status:in-progress`    | Work in progress.                        |
+| `status:inactive`       | Inactive PR or Issue. Likely to become a `candidate-for-closing` |
+| `status:pending`        | Issue is posted, but is not reviewed by maintainers. |
```
Seems useless. If maintainer review an issues it must drop a comment

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| :---------------------- | ---------------------------------------- |
+| `enhancement:code`      | Any improvements in **existing** code.   |
+| `enhancement:idea`      | Fresh ideas to enhance existing architecture, design.   |
+| `bug:needs-reproduction`| Bugs or reports that are very likely to be bugs. |
```
`bugs .. are very likely to be bugs`? Sound strange

Warchant replied:
*reports* that are very likely to be bugs

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `bug:needs-reproduction`| Bugs or reports that are very likely to be bugs. |
+| `bug:confirmed`         | Confirmed bug by maintainers. |
+| `question`              | Questions more than bug reports or feature requests - "How do I do X" |
+| `feature`               | Feature requests.                        |
```
Maybe `new feature`?

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `help-wanted`           | Maintainers ask for help to work on this issue. |
+| `accepted`              | Pull request is accepted and can be merged. |
+| `candidate-for-closing` | Outdated Issue/Pull Request and is candidate for closing. |
+| `challeging-task`       | Small but hard task.                     |
```
What is label for "big but hard task" then?

l4l replied:
Also remove `task` in the label name

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `accepted`              | Pull request is accepted and can be merged. |
+| `candidate-for-closing` | Outdated Issue/Pull Request and is candidate for closing. |
+| `challeging-task`       | Small but hard task.                     |
+| `needs-correction`      | Pull request or Issue that should be corrected by opener. |
```
s/opener/author/

l4l commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+| `status:wontfix`        | Core team has decided not to fix these issue for now. |
+
+
+### Contact Developers
```
Should be in iroha/README.md

l4l reviewed (changes requested):
As for CONTRIBUTING.md:
Maybe it's better to place a table with the necessary info and prefixes? (like bug, enhancement)

Also it can be splited it separate files

neewy commented:
corrections?

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+### Requirements
+
+* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
+* All new code must have [code coverage](https://docs.codecov.io/docs/about-code-coverage) above 70%.
```
It is written how it **should look**. So, new PRs should satisfy this requirement.

neewy replied:
It is not strict at all

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+<!-- You will not see HTML commented line in Pull Request body -->
+<!-- Optional sections may be omitted. Just remove them or write None -->
+
+### Description of the Change
```
Any PR is a **change**. "What is this pull request" does not sound like an English sentence.

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. -->
+<!-- Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->
+
+### Why Should This Be In Iroha?
```
We may have new contributors. These people should explain their motivation of introducing a change.

l4l replied:
Then maybe mark as optional?

neewy replied:
I agree. Sometimes the benefits and motivation are tightly coupled. Why this should be in Iroha might be a top item, while Benefits can be a subitem. I propose to delete it

neewy replied:
Agree with @l4l I mean

Warchant replied:
@neewy this template is first of all for **new contributors**

l4l replied:
Both new and old, actually

neewy replied:
Please make it optional or delete

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+
+<!-- Explain why this functionality should be in hyperledger/iroha -->
+
+### Benefits
```
New functionality is not needed if it does not give any benefits for the project.

neewy replied:
Please ignore this if you make Why Should This Be In Iroha? optional

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+
+<!-- What benefits will be realized by the code change? -->
+
+### Possible Drawbacks 
```
It is mandatory. If no drawbacks, PR author should explicitly mention this.

Warchant commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,22 +1,36 @@
+<!-- What are the possible side-effects or negative impacts of the code change? -->
+<!-- If no drawbacks, write None -->
+
+### Usage Examples *[optional]*
```
*Usage example* is a more broad term. 
It includes tests, checks, examples...

l4l replied:
Imo *usage* doesn't include tests/checks

neewy replied:
I agree. Please make this correction

Warchant commented on CONTRIBUTING.md:
````diff
@@ -1,44 +1,199 @@
+- When finished work, **rebase onto base branch** with 
+    ```bash
+    $ git fetch
+    $ git rebase -i <base-branch>
````
base branch may not be `origin`, if it is fork

l4l replied:
Yep, that's true and not really neat solution
Still it can lead to problems in contributors' prs, It's definetely need some explanation here, maybe more detailed info about the fetching, e.g:
```
git checkout <base-branch>
git pull <base-branch> -ff-only
git checkout <feature-branch>
git rebase -i <base-branch>
```

Warchant replied:
Below I mentioned step-by-step guide.

neewy replied:
No problem then

Warchant commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
 # Contributing guidelines
 
-To contribute, please follow this process:
+:star::tada: First off, thanks for taking the time to contribute! :tada::star:
```
First of all - прежде всего
First off - во-первых

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,199 @@
+
+- **Write tests** for new code.
+
+- Every pull request should be reviewed and **get at least two approvals**.
```
no-no, this is really important

l4l replied:
Bold text is for the emphasis and shouldn't be used frequently. Lot of bold text basically eliminates the emphasis.
I agree that marked text is really important, though it looks mostly like a mess

neewy commented on CONTRIBUTING.md:
```diff
@@ -1,44 +1,216 @@
+
+- Fill in [the required template](.github/PULL_REQUEST_TEMPLATE.md)
+
+- **Write tests** for new code.
```
can you mention coverage? like 70%

l4l replied:
May be just *reasonable* is enough?
For me it's better accept nice contributions with some meaningful tests that doesn't cover small and rare use-cases.

neewy reviewed (commented):
I have made some comments about small remaining issues. It would be great to fix them

l4l commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -25,9 +25,9 @@
 
 <!-- What are the possible side-effects or negative impacts of the code change? -->
-<!-- If no drawbacks, write None -->
+<!-- If no drawbacks, explicitly meantion this (write None) -->
```
me<s>a</s>ntion

neewy commented on .github_changelog_generator:
```diff
@@ -0,0 +1,7 @@
+unreleased=false
+issues-wo-labels=false
+pr-wo-labels=false
+exclude-labels=candidate-for-closing,status:in-progress,status:pending,status:active,status:inactive,needs-review,needs-correction,question,status:wontfix
```
please delete pending status

l4l commented on CONTRIBUTING.md:
```diff
@@ -91,7 +91,7 @@ Read our [C++ Style Guide](#c++-style-guide) and start with beginner-friendly is
 - Fill in [the required template](.github/PULL_REQUEST_TEMPLATE.md)
 
-- **Write tests** for new code.
+- **Write tests** for new code. Test coverage for new code be at least 70%.
```
..new code **should/must** be at..

neewy reviewed (changes requested):
Please implement these changes and I would be happy to have this merged :)

l4l reviewed (changes requested):
Amend the last commit with the 8607ebb on merging please

neewy commented:
I would be extra happy if you follow advice of @l4l and make html labels for issues
<a href="/hyperledger/iroha/labels/enhancement" class="label css-truncate-target linked-labelstyle-1d76db" title="enhancement">enhancement</a>

Warchant commented:
@neewy this does not work in github

l4l commented:
@neewy isaacs/github#305 is still unresolved :(

Warchant committed:
Documentation improvements

- enhance PR template
- add github changelog generator
- add code of conduct
- enhance contributing.md
- add apache 2 license full text

Signed-off-by: Bogdan Vaneev <warchantua@gmail.com>
